middleware/diesel/src/repo.rs

```diff
@@ -0,0 +1,172 @@
+    /// let repo = Repo::from_pool_builder(database_url,
+    ///     Pool::builder()
+    ///         .connection_timeout(Duration::from_secs(120))
+    ///         .max_size(100)
+    /// );
+    /// ```
+    pub fn from_pool_builder(
+        database_url: &str,
+        builder: r2d2::Builder<ConnectionManager<T>>,
+    ) -> Self {
+        let manager = ConnectionManager::new(database_url);
+        let connection_pool = builder
+            .build(manager)
+            .expect("could not initiate test db pool");
+        Repo { connection_pool }
+    }
+
+    /// Creates a repo for use in tests, where queries are executed
+    /// with an isolated test transaction and rolled back when
+    /// the connection is dropped. This allows tests to run in parallel
+    /// without impacting each other.
+    /// ```rust
+    /// # use diesel::sqlite::SqliteConnection;
+    ///
+    /// type Repo = gotham_middleware_diesel::Repo<SqliteConnection>;
+    /// let repo = Repo::with_test_transactions(":memory:");
+    /// ```
+    pub fn with_test_transactions(database_url: &str) -> Self {
+        let customizer = TestConnectionCustomizer {};
+        let builder = Pool::builder().connection_customizer(Box::new(customizer));
+        Self::from_pool_builder(database_url, builder)
+    }
+
+    /// Runs the given closure in a way that is safe for blocking IO to the
+    /// database without blocking the tokio reactor.
+    /// The closure will be passed a `Connection` from the pool to use.
+    pub fn run<F, R, E>(&self, f: F) -> impl Future<Item = R, Error = E>
+    where
+        F: FnOnce(PooledConnection<ConnectionManager<T>>) -> Result<R, E>
+            + Send
+            + std::marker::Unpin
+            + 'static,
+        T: Send + 'static,
+    {
+        let pool = self.connection_pool.clone();
+        // `tokio_threadpool::blocking` returns a `Poll` which can be converted into a future
+        // using `poll_fn`.
+        // `f.take()` allows the borrow checker to be sure `f` is not moved into the inner closure
+        // multiple times if `poll_fn` is called multple times.
+        let mut f = Some(f);
+        poll_fn(move || blocking(|| (f.take().unwrap())(pool.get().unwrap()))).then(
+            |future_result| match future_result {
+                Ok(query_result) => match query_result {
+                    Ok(result) => future::ok(result),
+                    Err(error) => future::err(error),
+                },
+                Err(e) => panic!(format!("Error running async database task: {:?}", e)),
+            },
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct TestConnectionCustomizer;
+
+impl<C, E> CustomizeConnection<C, E> for TestConnectionCustomizer
+where
+    C: diesel::connection::Connection,
+    E: std::error::Error + Sync + Send,
+{
+    fn on_acquire(&self, conn: &mut C) -> Result<(), E> {
+        match conn.begin_test_transaction() {
+            Ok(_) => Ok(()),
+            Err(_) => Ok(()),
```

Comment:
Does this maybe need a log or something? If it's only going to return `Ok(())`, then you can skip the match:

```
conn.begin_test_transaction();
Ok(())
```

Reply:
Yep - good idea to log something meaningful at least. updated.